packages/liquidator/src/proxyTransactionWrapper.js

```diff
@@ -0,0 +1,288 @@
+
+      // We can now fetch the reserves. At the same time, we can batch a few other required async calls.
+      const [
+        reserves,
+        token0,
+        contractPFC,
+        contractTokensOutstanding,
+        reserveTokenBalance,
+        collateralTokenBalance
+      ] = await Promise.all([
+        uniswapPair.getReserves(),
+        uniswapPair.token0(),
+        this.financialContract.methods.pfc().call(),
+        this.financialContract.methods.totalTokensOutstanding().call(),
+        this.reserveToken.methods.balanceOf(this.dsProxyManager.getDSProxyAddress()).call(),
+        this.collateralToken.methods.balanceOf(this.dsProxyManager.getDSProxyAddress()).call()
+      ]);
+
+      // Detect if the reserve currency is token0 or 1. This informs the order in the following computation.
+      const reserveToken0 = token0 == this.reserveToken._address;
+
+      const reserveIn = reserveToken0 ? reserves.reserve0 : reserves.reserve1;
+      const reserveOut = !reserveToken0 ? reserves.reserve0 : reserves.reserve1;
+
+      // Compute the maximum amount of collateral that can be purchased with all reserve currency.
+      const amountInWithFee = this.toBN(reserveTokenBalance).muln(997);
+      const numerator = amountInWithFee.mul(reserveOut);
+      const denominator = reserveIn.muln(1000).add(amountInWithFee);
+
+      const maxPurchasableCollateral = numerator.div(denominator);
+
+      // Compute how many synthetics could be minted, given the collateral we can buy with the reserve currency.
+      const gcr = this.toBN(contractPFC.rawValue)
+        .mul(this.toBN(this.toWei("1")))
+        .div(this.toBN(contractTokensOutstanding));
+
+      // Calculate the max mintable synthetics with the collateral. Use the sum of the max that can be purchased and any
+      // collateral the DSProxy already has in the computation.
+      const maxMintableSynthetics = maxPurchasableCollateral
+        .add(this.toBN(collateralTokenBalance))
+        .mul(this.toBN(this.toWei("1")))
+        .div(gcr);
+
+      // Finally, the effective synthetic balance is the sum of the max that can be minted using the collateral swapped
+      // from reserve and the current synthetics within the DSProxy account.
+      return maxMintableSynthetics.add(this.toBN(syntheticTokenBallance));
+    }
+  }
+
+  // Main entry point for submitting a liquidation. If the bot is not using a DSProxy then simply send a normal EOA tx.
+  // If the bot is using a DSProxy then route the tx via it.
+  async submitLiquidationTransaction(liquidationArgs) {
+    // If the liquidator is not using a DSProxy, use the old method of liquidating
+    if (!this.isUsingDsProxyToLiquidate) return await this._executeLiquidationWithoutDsProxy(liquidationArgs);
+    else return await this._executeLiquidationWithDsProxy(liquidationArgs);
+  }
+
+  async _executeLiquidationWithoutDsProxy(liquidationArgs) {
```

Comment:
Maybe this isn't the PR to do this but we should probably refactor the `ynatm` retry logic into `common/src/ContractUtils.js:runTransaction()`, which currently manages transactions for the `funding-rate-proposer` and `optimistic-oracle-proposer`. This would remove the `estimateGas-send` code below.

Reply:
I'll work on this in a spearate PR